File: loadFiles.ts
import { parse } from "@std/csv/parse";
import type { Chall, Challs } from "./types.ts";
import { CsvStream } from "https://deno.land/std@0.170.0/encoding/csv/stream.ts";

export async function loadChalls(filePath: string): Promise<Challs> {
  const file = await Deno.readTextFile(filePath);

  const parsedData = parse(file, {
    skipFirstRow: true,
  });

  const challs: Challs = {};

  for (const row of parsedData) {
    const id = row.id;
    const genre = row.category;
    const name = row.name;
    try {
      const score = Number(row.value);
      const solver = Number(0);
      const chall: Chall = {
        name,
        genre,
        score,
        solver,
      };
      challs[id] = chall;
    } catch (e) {
      console.log(`Error: ${e}`);
    }
  }
  return challs;
}

export async function loadTeams(filePath: string): Promise<Set<string>> {
  const file = await Deno.readTextFile(filePath);
  const parsedData = parse(file, {
    skipFirstRow: true,
  });

  const ignoredTeams = new Set<string>();

  for (const row of parsedData) {
    const id = row.id;
    const hidden = row.hidden;
    const banned = row.banned;

    if (hidden === "True" || banned === "True") {
      ignoredTeams.add(id);
    }
  }
  return ignoredTeams;
}

// solvesはファイルサイズが大きいため、ストリーミング処理を行う
export async function loadSolves(
  _filePath: string,
  challs: Challs,
  ignoredTeams: Set<string> = new Set(),
) {
  let filePath = _filePath;
  if (
    !filePath.startsWith("./") &&
    !filePath.startsWith("../") &&
    !filePath.startsWith("/")
  ) {
    filePath = `./${filePath}`;
  }
  const path = new URL(import.meta.resolve(filePath));
  const { readable } = await Deno.open(path);

  const data = readable
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(new CsvStream());

  let passedHeader = false;
  for await (const line of data) {
    if (!passedHeader) {
      passedHeader = true;
      continue;
    }
    const challId = line[0];
    const teamId = line[2];

    if (ignoredTeams.has(teamId)) {
      continue;
    }

    if (challs[challId]) {
      challs[challId].solver++;
    }
  }
  return challs;
}

File: load_files.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"strconv"
)

func loadChalls(filePath string) (Challs, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	header, err := reader.Read()
	if err != nil {
		return nil, err
	}

	colIdx := make(map[string]int)
	for i, h := range header {
		colIdx[h] = i
	}

	challs := make(Challs)
	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}

		id := record[colIdx["id"]]
		name := record[colIdx["name"]]
		genre := record[colIdx["category"]]
		score, err := strconv.Atoi(record[colIdx["value"]])
		if err != nil {
			fmt.Printf("Error: %v\n", err)
			continue
		}

		challs[id] = &Chall{
			Name:   name,
			Genre:  genre,
			Score:  score,
			Solver: 0,
		}
	}
	return challs, nil
}

func loadTeams(filePath string) (map[string]bool, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	header, err := reader.Read()
	if err != nil {
		return nil, err
	}

	colIdx := make(map[string]int)
	for i, h := range header {
		colIdx[h] = i
	}

	ignoredTeams := make(map[string]bool)
	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}

		id := record[colIdx["id"]]
		hidden := record[colIdx["hidden"]]
		banned := record[colIdx["banned"]]

		if hidden == "True" || banned == "True" {
			ignoredTeams[id] = true
		}
	}
	return ignoredTeams, nil
}

// loadSolves はsolves CSVを読み込み、各チャレンジのsolver数を更新する。
// solvesはファイルサイズが大きいため、ストリーミング処理を行う。
func loadSolves(filePath string, challs Challs, ignoredTeams map[string]bool) (Challs, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	// Skip header
	if _, err := reader.Read(); err != nil {
		return nil, err
	}

	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}

		challID := record[0]
		teamID := record[2]

		if ignoredTeams[teamID] {
			continue
		}

		if chall, ok := challs[challID]; ok {
			chall.Solver++
		}
	}
	return challs, nil
}
